File: SelfPlay.py
import numpy as np
from torch.utils.data import DataLoader

import Logger
import Network
import Utils
from Game import Game
from MCTS import Node, MCTS
from SampleSet import SampleSet
import time

import gc


def sub_log(num_processes, i_numGames, text):
    Logger.info(f"进程{num_processes} 第{i_numGames}局 - {text} ")


def selfPlay(boardSize, tie_mu, numGames, num_processes,
             numSimulations, temperatureDefault, explorationFactor):
    onnx_model = Network.load_onnx_model("model/model_latest_fp16.onnx")

    if onnx_model is None:
        return []

    training_data = []

    for i_numGames in range(numGames):
        game = Game(board_size=boardSize, device=Utils.getDevice(), tie_mu=tie_mu)
        root = Node()
        mcts = MCTS(model=onnx_model, exploration_constant=explorationFactor)
        mcts.root = root

        step = 0

        actions = []
        while True:
            start_time = time.time()

            step += 1

            needSearchCount = numSimulations - mcts.root.visits

            sub_log(num_processes, i_numGames, f"计划模拟次数 {needSearchCount},  第 {step} 步")
            mcts.search(game, needSearchCount)

            # 步骤 1: 提取 visit 的数值
            # 步骤 2: 归一化得到概率分布
            # 步骤 3: 以这个分布的概率，随机取一个对象的索引

            visit_values = np.array([obj.visits for obj in mcts.root.children])
            visit_sum = visit_values.sum()

            probabilities = visit_values / visit_sum

            random_index = np.random.choice(len(mcts.root.children), p=probabilities)
            node = mcts.root.children[random_index]

            # 打印决策
            probs_matrix = np.zeros((game.board_size, game.board_size))
            for child in mcts.root.children:
                if child.move[0] >= 0:
                    probs_matrix[child.move[0]][child.move[1]] = child.visits / visit_sum
            # print(probs_matrix)

            actions.append((Network.get_state(game), game.current_player, probs_matrix))

            end_time = time.time()
            sub_log(num_processes, i_numGames,
                    f"玩家 {game.current_player}, 落子  {node.move},  访问次数 {node.visits} "
                    f", 速度 {round(needSearchCount / (end_time - start_time), 1)} 次/秒")
            game.make_move(node.move[0], node.move[1])

            if game.end_game_check():
                break
            # print(actions)
            game.render()

            # 重置跟节点，并清理其他分支内存
            node.parent = None
            mcts.root.children.remove(node)

            mcts.root.clear()
            del mcts.root

            mcts.root = node

            ### 强制进行垃圾回收
            gc.collect()

        winner = game.calculate_winner()
        sub_log(num_processes, i_numGames, f"本局胜方 玩家 {winner}")

        game.render()

        for (state, player, probs) in actions:
            value = 0
            if winner == player:
                value = 1.0
            if winner == 3 - player:
                value = -1.0
            training_data.append((state, probs, value))

    return training_data


if __name__ == "__main__":
    model, _ = Network.get_model(Utils.getDevice(), 1e-3)
    training_data = selfPlay(9, 3, 1, 0, 10, 1, 3)
    print(training_data)
    sample_set = SampleSet(training_data)
    dataloader = DataLoader(sample_set, batch_size=32, shuffle=True)
    # 训练循环
    running_loss = 0.0
    for batch_data in dataloader:
        states = batch_data[0].float().to(Utils.getDevice())
        mcts_probs = batch_data[1].float().to(Utils.getDevice())
        values = batch_data[2].float().to(Utils.getDevice())
        values = values.unsqueeze(dim=0)
        print(values.shape)

File: SelfPlay.cpp

#include "SelfPlay.h"

#include "ConfigReader.h"
#include "Utils.h"

// 创建一个随机数生成器
std::random_device rd;
std::mt19937 gen(rd());


void printGame(Model &model, Game &game, Point move, double rate,
               float temperature, vector<float> probs) {
    game.render();

    auto state = Model::get_state(game);
    auto [value, prob_metrix] = model.evaluate_state(state);

    std::string pic = ((3 - game.currentPlayer) == 1) ? "x" : "o";
    cout << pic << " move is " << move.x << ","
            << move.y
            << " on rate " << round(rate * 1000) / 1000
            << " temperature " << temperature
            << " value " << -value
            << endl;


    std::cout << std::fixed << std::setprecision(3);
    for (int i = 0; i < probs.size(); i++) {
        cout << probs[i] << " ";
        if (i % game.boardSize == game.boardSize - 1) {
            cout << endl;
        }
    }
}


tuple<float, Point, float> getNextMove(int step, float temperatureDefault,vector<float>& move_probs,
                                       vector<Point>& moves, MonteCarloTree& mcts)
{
    //按温度决策
    float temperature;
    Point move;
    float rate;

    if (step < stoi(ConfigReader::get("temperatureDownBeginStep")))
    {
        //前n步，温度>0
        temperature = temperatureDefault;
        std::discrete_distribution<int> distribution(move_probs.begin(), move_probs.end());
        int index = distribution(gen);
        move = moves[index];
        rate = move_probs[index];
    }
    else
    {
        //温度为0
        temperature = 0;
        move = mcts.get_max_visit_move();
        rate = 1;
    }
    return tuple(temperature, move, rate);
}

std::vector<std::tuple<vector<vector<vector<float> > >, std::vector<float>, std::vector<float> > > selfPlay(
    int shard,
    int boardSize,
    float tieMu,
    int numGames,
    int numSimulations,
    float temperatureDefault,
    float explorationFactor,
    Model &model
) {
    MonteCarloTree mcts = MonteCarloTree(&model, explorationFactor, true);
    std::vector<std::tuple<vector<vector<vector<float> > >, std::vector<float>, std::vector<float> > > training_data;

    for (int i = 0; i < numGames; i++) {
        Game game(boardSize, tieMu);
        std::vector<std::tuple<vector<vector<vector<float> > >, int, std::vector<float> > > game_data;

        int step = 0;
        while (!game.endGameCheck()) {
            Node node;
            //开始mcts预测
            auto startTime = getSystemTime();
            int simiNum = numSimulations;
            mcts.search(game, &node, simiNum);
            if (simiNum > 0) {
                cout << "======== " << shard << "-" << i << " =======" << endl << "search cost " << getSystemTime() -
                        startTime << " ms, simi num " << simiNum << ", "
                        << "per simi " << (getSystemTime() - startTime) / (float) simiNum << " ms" << endl;
            }

            auto [moves, moves_probs] = mcts.get_action_probabilities();
            //决策下一步
            auto [temperature, move, rate] = getNextMove(step, temperatureDefault, moves_probs, moves, mcts);
            std::tie(moves, moves_probs) = mcts.get_action_probabilities();

            // 随机选择
            std::discrete_distribution<int> distribution(moves_probs.begin(),
                                                         moves_probs.end());

            // 构造矩阵
            vector<float> probs_matrix(game.boardSize * game.boardSize, 0);

            if (!moves[0].isNull()) {
                for (int k = 0; k < moves.size(); k++) {
                    auto p = moves[k];
                    probs_matrix[game.getMoveIndex(p.x, p.y)] = moves_probs[k];
                }
            }
            game_data.emplace_back(Model::get_state(game), game.currentPlayer, probs_matrix);

            // cout << "bannedMoves: ";
            // for (auto move: game.bannedMoves) {
            //     cout << "(" << move.x << "," << move.y << ") ";
            // }
            // cout << endl;
            //
            // cout << "eatMoves: ";
            // for (auto [move, gourp]: game.eatMoves) {
            //     cout << "(" << move.x << "," << move.y << ") ";
            // }
            // cout << endl;

            game.makeMove(move.x, move.y);

            printGame(model, game, move, rate, temperature, probs_matrix);
            step++;
        }

        auto winner = game.calculateWinner();
        for (const auto &[state, player, mcts_probs]: game_data) {
            float value = 0;
            if (winner == player) {
                value = 1;
            }
            if (winner == 3 - player) {
                value = -1;
            }
            training_data.emplace_back(state, mcts_probs, std::vector<float>{value});
        }
        auto [blackScore,whiteScore] = game.calculateScore();
        cout << "winner is " << ((winner == 1) ? "x" : ((winner == 2) ? "o" : " nobody"))
                << " score " << blackScore << ":" << whiteScore << endl;
    }
    return training_data;
}

void recordSelfPlay(
    int boardSize,
    float tieMu,
    int numGames,
    int numSimulations,
    float temperatureDefault,
    float explorationFactor,
    int shard,
    string *modelPath,
    string *coreType) {
    try {
        Model model;
        model.init(*modelPath, *coreType);

        // 创建文件流对象
        std::ofstream file("record/data_" + to_string(shard) + ".txt");

        if (file.is_open()) {
            auto data = selfPlay(shard, boardSize, tieMu, numGames, numSimulations,
                                 temperatureDefault, explorationFactor, model);
            file << data.size() << endl;
            std::cout << "data count " << data.size() << endl;
            for (auto &item: data) {
                auto state = get<0>(item);

                // 获取张量的维度
                int64_t dim0 = state.size();
                int64_t dim1 = state[0].size();
                int64_t dim2 = state[0][0].size();

                file << dim0 << " " << dim1 << " " << dim2 << endl;
                // 遍历张量并打印数值
                for (int64_t i = 0; i < dim0; ++i) {
                    for (int64_t j = 0; j < dim1; ++j) {
                        for (int64_t k = 0; k < dim2; ++k) {
                            file << state[i][j][k] << " ";
                        }
                        file << endl;
                    }
                }

                vector<float> mctsProbList = get<1>(item);
                file << mctsProbList.size() << endl;
                for (auto f: mctsProbList) {
                    file << f << " ";
                }
                file << endl;

                vector<float> valueList = get<2>(item);
                file << valueList.size() << endl;
                for (auto f: valueList) {
                    file << f << " ";
                }
                file << endl;
            }

            // 关闭文件
            file.close();
            std::cout << "Data has been written to file." << std::endl;
        } else {
            std::cerr << "Failed to open file." << std::endl;
        }
    } catch (const std::exception &e) {
        // 打印异常信息
        std::cerr << "Exception in shard " << shard << ": " << e.what() << std::endl;
        std::cout << "Exception in shard " << shard << ": " << e.what() << std::endl;
    }
}
